File: src/common/main/java/hauveli/fishcasting/hexcasting/recipe/brainsweep/BorderTypeEntityTypeIngredient.java
package hauveli.fishcasting.hexcasting.recipe.brainsweep;

import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.BrainsweepeeIngredientType;
import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.EntityTypeIngredient;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import io.netty.buffer.ByteBuf;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.StringRepresentable;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.border.WorldBorder;
import net.minecraft.world.phys.Vec3;

import java.util.Objects;

public class BorderTypeEntityTypeIngredient extends EntityTypeIngredient {
    public final BorderType upDownBorder;
    public final BorderType northSouthBorder;
    public final BorderType westEastBorder;

    /*
    @Override
    public Entity exampleEntity(Level level) {
        return this.entityType.create(level);
    }
     */

    // public final EntityType<?> entityType ;

    public BorderTypeEntityTypeIngredient(EntityType<?> entityType,
                                          BorderType upDownBorder,
                                          BorderType northSouthBorder,
                                          BorderType westEastBorder) {
        super(entityType);
        this.upDownBorder = upDownBorder;
        this.northSouthBorder = northSouthBorder;
        this.westEastBorder = westEastBorder;
    }

    public BorderType getUpDownBorder() {
        return this.upDownBorder;
    }
    public BorderType getNorthSouthBorder() {
        return this.northSouthBorder;
    }
    public BorderType getWestEastBorder() {
        return this.westEastBorder;
    }

    enum BorderType implements StringRepresentable {
        POSITIVE("positive"),
        NEGATIVE("negative"),
        ANY("any");

        private final String name;

        BorderType(String name) {
            this.name = name;
        }

        @Override
        public String getSerializedName() {
            return this.name;
        }

        public static final Codec<BorderType> CODEC = StringRepresentable.fromEnum(BorderType::values);
        public static final StreamCodec<ByteBuf, BorderType> STREAM_CODEC = ByteBufCodecs.STRING_UTF8.map(BorderType::valueOf, BorderType::name); // should I use UTF8? I didn't see other String options that made sense...
    }

    private boolean positionAtBorder(BorderType borderType, double coordinateLimit, double coordinateOnAxis) {
        switch (borderType) {
            case POSITIVE -> {
                return coordinateLimit <= coordinateOnAxis;
            }
            case NEGATIVE -> {
                return coordinateLimit >= coordinateOnAxis;
            }
            case ANY -> {
                return true;
            }
        }
        return false; // unreachable, I think, but the compiler goes waah waah waah
    }

    private boolean notAtBorder(BorderType borderType, double minimum, double maximum, double position) {
        if (borderType == BorderType.ANY) {
            return false;
        }
        double limit = minimum;
        if (borderType != BorderType.NEGATIVE) {
            limit = maximum;
        }
        return !positionAtBorder(borderType, limit, position);
    }

    private boolean entityAtBorder(Entity entity) {
        Vec3 pos = entity.position();
        if (notAtBorder(getUpDownBorder(),
                entity.level().getMinBuildHeight(),
                entity.level().getMaxBuildHeight(),
                pos.y)) {
            return false;
        }
        WorldBorder worldBorder = entity.level().getWorldBorder();
        if (notAtBorder(getNorthSouthBorder(),
                worldBorder.getMinZ(),
                worldBorder.getMaxZ(),
                pos.z)) {
            return false;
        }
        if (notAtBorder(getWestEastBorder(),
                worldBorder.getMinX(),
                worldBorder.getMaxX(),
                pos.x)) {
            return false;
        }
        return true;
    }

    @Override
    public boolean test(Entity entity, ServerLevel level) {
        // entity types are singletons
        return entity.getType() == this.entityType
                && this.entityAtBorder(entity);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        BorderTypeEntityTypeIngredient that = (BorderTypeEntityTypeIngredient) o;
        return Objects.equals(entityType, that.entityType);
    }

    public static class Type implements BrainsweepeeIngredientType<BorderTypeEntityTypeIngredient> {
        public static final MapCodec<BorderTypeEntityTypeIngredient> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
                BuiltInRegistries.ENTITY_TYPE.byNameCodec().fieldOf("entityType").forGetter(BorderTypeEntityTypeIngredient::getEntityType),
                BorderType.CODEC.fieldOf("upDownBorder").forGetter(BorderTypeEntityTypeIngredient::getUpDownBorder),
                BorderType.CODEC.fieldOf("northSouthBorder").forGetter(BorderTypeEntityTypeIngredient::getNorthSouthBorder),
                BorderType.CODEC.fieldOf("westEastBorder").forGetter(BorderTypeEntityTypeIngredient::getWestEastBorder)
        ).apply(instance, BorderTypeEntityTypeIngredient::new));
        public static final StreamCodec<RegistryFriendlyByteBuf, BorderTypeEntityTypeIngredient> STREAM_CODEC = StreamCodec.composite(
                ByteBufCodecs.registry(Registries.ENTITY_TYPE), BorderTypeEntityTypeIngredient::getEntityType,
                BorderType.STREAM_CODEC, BorderTypeEntityTypeIngredient::getUpDownBorder,
                BorderType.STREAM_CODEC, BorderTypeEntityTypeIngredient::getNorthSouthBorder,
                BorderType.STREAM_CODEC, BorderTypeEntityTypeIngredient::getWestEastBorder,
                BorderTypeEntityTypeIngredient::new
        );

        @Override
        public MapCodec<BorderTypeEntityTypeIngredient> codec() {
            return CODEC;
        }

        @Override
        public StreamCodec<RegistryFriendlyByteBuf, BorderTypeEntityTypeIngredient> streamCodec() {
            return STREAM_CODEC;
        }
    }
}

File: src/common/main/kotlin/hauveli/fishcasting/casting/recipe/brainsweep/BorderTypeEntityTypeIngredient.kt
package hauveli.fishcasting.casting.recipe.brainsweep

import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.BrainsweepeeIngredientType
import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.EntityTypeIngredient
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import io.netty.buffer.ByteBuf
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.StringRepresentable
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import java.util.function.Function


class BorderTypeEntityTypeIngredient(
    entityType: EntityType<*>?,
    val upDownBorder: BorderType?,
    val northSouthBorder: BorderType?,
    val westEastBorder: BorderType?
) : EntityTypeIngredient(entityType) {
    enum class BorderType(private val nameId: String) : StringRepresentable {
        POSITIVE("positive"),
        NEGATIVE("negative"),
        ANY("any");

        override fun getSerializedName(): String {
            return this.nameId
        }

        companion object {
            val CODEC: Codec<BorderType?> =
                StringRepresentable.fromEnum({ entries.toTypedArray() })
            val STREAM_CODEC: StreamCodec<ByteBuf?, BorderType?> = ByteBufCodecs.STRING_UTF8.map<BorderType?>(
                Function { name: String? -> BorderType.valueOf(name!!) },
                Function { obj: BorderType? -> obj!!.name }) // should I use UTF8? I didn't see other String options that made sense...
        }
    }

    private fun positionAtBorder(borderType: BorderType?, coordinateLimit: Double, coordinateOnAxis: Double): Boolean {
        return when (borderType) {
            BorderType.POSITIVE -> {
                coordinateLimit <= coordinateOnAxis
            }

            BorderType.NEGATIVE -> {
                coordinateLimit >= coordinateOnAxis
            }

            BorderType.ANY -> {
                true
            }

            else -> {
                false // this is unreachable at runtime...
            }
        }
        //return false // unreachable, I think, but the compiler goes waah waah waah
    }

    private fun notAtBorder(borderType: BorderType?, minimum: Double, maximum: Double, position: Double): Boolean {
        if (borderType == BorderType.ANY) {
            return false
        }
        var limit = minimum
        if (borderType != BorderType.NEGATIVE) {
            limit = maximum
        }
        return !positionAtBorder(borderType, limit, position)
    }

    private fun entityAtBorder(entity: Entity): Boolean {
        val pos = entity.position()
        if (notAtBorder(
                this.upDownBorder,
                entity.level().minBuildHeight.toDouble(),
                entity.level().maxBuildHeight.toDouble(),
                pos.y
            )
        ) {
            return false
        }
        val worldBorder = entity.level().worldBorder
        if (notAtBorder(
                this.northSouthBorder,
                worldBorder.minZ,
                worldBorder.maxZ,
                pos.z
            )
        ) {
            return false
        }
        if (notAtBorder(
                this.westEastBorder,
                worldBorder.minX,
                worldBorder.maxX,
                pos.x
            )
        ) {
            return false
        }
        return true
    }

    override fun test(entity: Entity, level: ServerLevel?): Boolean {
        // entity types are singletons
        return entity.type === this.entityType
                && this.entityAtBorder(entity)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as BorderTypeEntityTypeIngredient
        return entityType == that.entityType
    }

    class Type : BrainsweepeeIngredientType<BorderTypeEntityTypeIngredient?> {
        override fun codec(): MapCodec<BorderTypeEntityTypeIngredient?>? {
            return CODEC
        }

        override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf?, BorderTypeEntityTypeIngredient?> {
            return STREAM_CODEC
        }

        companion object {
            val CODEC: MapCodec<BorderTypeEntityTypeIngredient?>? =
                RecordCodecBuilder.mapCodec<BorderTypeEntityTypeIngredient?>(
                    Function { instance: RecordCodecBuilder.Instance<BorderTypeEntityTypeIngredient?>? ->
                        instance!!.group<EntityType<*>?, BorderType?, BorderType?, BorderType?>(
                            BuiltInRegistries.ENTITY_TYPE.byNameCodec().fieldOf("entityType")
                                .forGetter<BorderTypeEntityTypeIngredient?>(
                                    { obj: BorderTypeEntityTypeIngredient? -> obj!!.getEntityType() }),
                            BorderType.CODEC.fieldOf("upDownBorder")
                                .forGetter<BorderTypeEntityTypeIngredient?>({ obj: BorderTypeEntityTypeIngredient? -> obj!!.upDownBorder }),
                            BorderType.CODEC.fieldOf("northSouthBorder")
                                .forGetter<BorderTypeEntityTypeIngredient?>({ obj: BorderTypeEntityTypeIngredient? -> obj!!.northSouthBorder }),
                            BorderType.CODEC.fieldOf("westEastBorder")
                                .forGetter<BorderTypeEntityTypeIngredient?>({ obj: BorderTypeEntityTypeIngredient? -> obj!!.westEastBorder })
                        ).apply<BorderTypeEntityTypeIngredient?>(
                            instance,
                            { entityType: EntityType<*>?, upDownBorder: BorderType?, northSouthBorder: BorderType?, westEastBorder: BorderType? ->
                                BorderTypeEntityTypeIngredient(
                                    entityType,
                                    upDownBorder,
                                    northSouthBorder,
                                    westEastBorder
                                )
                            })
                    })
            val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf?, BorderTypeEntityTypeIngredient?> =
                StreamCodec.composite<RegistryFriendlyByteBuf?, BorderTypeEntityTypeIngredient?, EntityType<*>?, BorderType?, BorderType?, BorderType?>(
                    ByteBufCodecs.registry(Registries.ENTITY_TYPE),
                    { obj: BorderTypeEntityTypeIngredient? -> obj!!.getEntityType() },
                    BorderType.STREAM_CODEC,
                    { obj: BorderTypeEntityTypeIngredient? -> obj!!.upDownBorder },
                    BorderType.STREAM_CODEC,
                    { obj: BorderTypeEntityTypeIngredient? -> obj!!.northSouthBorder },
                    BorderType.STREAM_CODEC,
                    { obj: BorderTypeEntityTypeIngredient? -> obj!!.westEastBorder },
                    { entityType: EntityType<*>?, upDownBorder: BorderType?, northSouthBorder: BorderType?, westEastBorder: BorderType? ->
                        BorderTypeEntityTypeIngredient(
                            entityType,
                            upDownBorder,
                            northSouthBorder,
                            westEastBorder
                        )
                    }
                )
        }
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + upDownBorder.hashCode()
        result = 31 * result + northSouthBorder.hashCode()
        result = 31 * result + westEastBorder.hashCode()
        return result
    }
}